exchange/peermgr.go

```diff
@@ -63,11 +64,86 @@ func NewPeerMgr(h host.Host, regions []Region) *PeerMgr {
 			}
 		},
 	})
+
 	return pm
 }
 
+func (pm *PeerMgr) Run(ctx context.Context) error {
+	pm.h.SetStreamHandler(HeyProtocol, pm.handleStream)
+
+	sub, err := pm.h.EventBus().Subscribe(new(event.EvtPeerIdentificationCompleted), eventbus.BufSize(1024))
+	if err != nil {
+		return fmt.Errorf("failed to subscribe to event bus: %w", err)
+	}
+
+	go func() {
+		for evt := range sub.Out() {
+			pic := evt.(event.EvtPeerIdentificationCompleted)
+			go func() {
+				if err := pm.sendHey(ctx, pic.Peer); err != nil {
+					return
+				}
+			}()
+		}
+	}()
+	return nil
+}
+
+// Peers returns n active peers for a given list of regions and peers to ignore
+func (pm *PeerMgr) Peers(n int, rl []Region, ignore map[peer.ID]bool) []peer.ID {
+	var peers []peer.ID
+	if n == 0 {
+		return peers
+	}
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+	for _, r := range rl {
+		for p, v := range pm.peers {
+			if ignore[p] {
+				continue
+			}
+			for _, rc := range v.Regions {
+				if rc == r.Code {
+					peers = append(peers, p)
+				}
+			}
+			// Check if we have enough peers and return
+			if len(peers) == n {
+				return peers
+			}
+		}
+	}
+	return peers
+}
+
+// handleStream is the multistream handler for the Hey protocol, it reads a Hey message and handles it
+func (pm *PeerMgr) handleStream(s network.Stream) {
+	var hmsg Hey
+	if err := cborutil.ReadCborRPC(s, &hmsg); err != nil {
+		connErr := s.Conn().Close()
+		if connErr != nil {
+			log.Error().Err(connErr).Msg("could not close stream connection")
+		}
+		log.Error().Err(err).Msg("failed to read CBOR Hey msg")
+		return
+	}
+
+	pm.receive(s.Conn().RemotePeer(), hmsg)
+
+	// We send back the seed to measure roundrip time
+	go func() {
+		defer s.Close()
+		buf := make([]byte, 32)
+
+		_, err := s.Write(buf)
+		if err != nil {
+			log.Error().Err(err).Msg("could not write bytes")
+		}
+	}()
+}
+
 // Receive a new greeting from peer
-func (pm *PeerMgr) Receive(p peer.ID, h Hey) {
+func (pm *PeerMgr) receive(p peer.ID, h Hey) {
```

Comment:
```suggestion
func (pm *PeerMgr) receiveHey(p peer.ID, h Hey) {
```
Might be more consistent

Reply:
Hum, how about handleHey instead? Because we already have a getHey

Reply:
receiveHey isn't the same as getHey though

Reply:
ah yes true but I though it was confusing